---
output:
  md_document:
    variant: markdown_github
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# Phased Or Unphased LD (pould) 
v0.10.5.9000 (February 21, 2020)

The *pould* package calculates four linkage disequilibrium (LD) statistics -- *D<sup>'</sup>*, *W<sub>n</sub>* and the two conditional asymmetric LD (cALD) measures, *W<sub>A/B</sub>* and *W<sub>B/A</sub>* -- for genotype data from pairs of genetic loci, and can treat these data as either phased or unphased for these calculations. In addition, *pould* includes *LDWrap()*, a wrapper function that parses genotype data in BIGDAWG/PyPop input format or haplotype data in HaplObserve output format, *LD.sign.test()*, which applies a sign test to LD values for phased and unphased haplotypes generated by *LDWrap()* for a given dataset, and *LD.heat.map()*, which generates PNG-formatted heat-map plots for each LD measure. 

For examples of the application of the *pould* package, see:<br>
Osoegawa et al. [Hum Immunol. In Press, 2019a](https://doi.org/10.1016/j.humimm.2019.01.010).<br>
Osoegawa et al. [Hum Immunol. In Press, 2019b](https://doi.org/10.1016/j.humimm.2019.05.018).

For more information about cALD, see: Thomson G, Single RM. **Conditional asymmetric linkage disequilibrium (ALD): extending the biallelic r2 measure.** [Genetics. 2014;198(1):321-31](https://doi.org/10.1534/genetics.114.165266).

The *pould* package can be installed from GitHub using the R *devtools* package --  `devtools::install_github("IHIW/pould/pould", build_vignettes = TRUE)`.

**Note**: When installing *pould* from GitHub in a Windows environment, the following warning message may appear on Windows systems that do not have Rtools v3.5 installed:
`In untar2(tarfile, files, list, exdir) :
  skipping pax global extended headers
`
This warning does not impact the function of the package. Installing [Rtools v3.5](https://cran.r-project.org/bin/windows/Rtools/) will prevent these warnings.

## Example
In addiiton to simply calculating LD values, *pould* can be used to compare LD for phased and unphased versions of the same dataset, e.g., to examine the extent to which phasing via segregation analysis impacts LD relative to phasing estimation via the expectation-maximization (EM) algorithm. In the example below, DRB1 and DQB1 genotype data were extracted from six-locus haplotypes that had been phased using the EM method ([Mack et al. Genes Immun. 2018](https://doi.org/10.1038/s41435-017-0006-8)). In the first application of *cALD()*, that phasing information is ignored, and the EM algorithm is applied to estimate haplotypes. In the second application of *cALD()*, the original six-locus phasing information is retained.

By comparing the resulting LD values, it becomes clear that LD is uniformly *lower* for the pre-phased DRB1~DQB1 haplotypes than for the *de novo* EM estimated haplotypes. This suggests that the EM algorithm may not be accurately estimating haplotypes low-frequency (counts < 4) haplotypes for individual locus pairs during multi-locus haplotype estimation, as the number of EM estimated haplotypes evaluated (53) is considerably lower than the number of pre-phased haplotypes evaluaded (106). 

```{r example}
## Comparing LD values for haplotypes generated by the EM algorithm (default = unphased) to LD values for haplotypes for which phased is known.
library("pould")
data(drb1.dqb1.demo)
cALD(drb1.dqb1.demo,inPhase=FALSE)

cALD(drb1.dqb1.demo,inPhase=TRUE)
```
